larq_compute_engine/core/indirect_bgemm/kernel_8x4x1_aarch64.h

```diff
@@ -0,0 +1,630 @@
+ * the first iteration of the next accumulation loop.
+ */
+
+// Float output transform
+inline void OutputTransformAndLoadNextAndStore(
+    const std::int32_t c_out_index, const std::int32_t channels_out,
+    uint16x8_t accumulators[4], int32x4_t weights[2], int32x4_t activations[4],
+    const bconv2d::OutputTransform<float>& output_transform,
+    const TBitpacked* weights_ptr, TBitpacked** indirection_buffer,
+    float*& output_ptr_0, float*& output_ptr_1, float*& output_ptr_2,
+    float*& output_ptr_3) {
+  ruy::profiler::ScopeLabel label("Float output transform and store");
+
+  // Declare result registers.
+  float32x4x2_t results[4];
+
+  asm("ldp x0, x1, [%[indirection_buffer]]\n\t"
+
+      // Use "unsigned shift left long" instructions to perform the
+      // back-transformation left-shift and extend the result to int32.
+      "ld1r {v0.4s}, [%[clamp_min_addr]]\n\t"
+      "ushll %[result_00].4s, %[accum_0].4h, #1\n\t"
+      "ushll %[result_10].4s, %[accum_1].4h, #1\n\t"
+      "ld1r {v1.4s}, [%[clamp_max_addr]]\n\t"
+      "ushll %[result_20].4s, %[accum_2].4h, #1\n\t"
+      "ushll %[result_30].4s, %[accum_3].4h, #1\n\t"
+      "ldr q2, [%[multiplier_addr]]\n\t"
+      "ushll2 %[result_31].4s, %[accum_3].8h, #1\n\t"
+      "ushll2 %[result_21].4s, %[accum_2].8h, #1\n\t"
+      "ldr q3, [%[multiplier_addr], #16]\n\t"
+      "ushll2 %[result_11].4s, %[accum_1].8h, #1\n\t"
+      "ushll2 %[result_01].4s, %[accum_0].8h, #1\n\t"
+
+      "ldp x2, x3, [%[indirection_buffer], #16]\n\t"
+
+      // Perform clamping
+      "ldr q4, [%[bias_addr]]\n\t"
+      "smax %[result_30].4s, %[result_30].4s, v0.4s\n\t"
+      "smax %[result_31].4s, %[result_31].4s, v0.4s\n\t"
+      "smax %[result_20].4s, %[result_20].4s, v0.4s\n\t"
+      "smax %[result_21].4s, %[result_21].4s, v0.4s\n\t"
+      "ldr q5, [%[bias_addr], #16]\n\t"
+      "smax %[result_10].4s, %[result_10].4s, v0.4s\n\t"
+      "smax %[result_11].4s, %[result_11].4s, v0.4s\n\t"
+      "smax %[result_00].4s, %[result_00].4s, v0.4s\n\t"
+      "smax %[result_01].4s, %[result_01].4s, v0.4s\n\t"
+      "ld1r {%[a_0].4s}, [x0]\n\t"
+      "smin %[result_30].4s, %[result_30].4s, v1.4s\n\t"
+      "smin %[result_31].4s, %[result_31].4s, v1.4s\n\t"
+      "smin %[result_20].4s, %[result_20].4s, v1.4s\n\t"
+      "smin %[result_21].4s, %[result_21].4s, v1.4s\n\t"
+      "ld1r {%[a_1].4s}, [x1]\n\t"
+      "smin %[result_10].4s, %[result_10].4s, v1.4s\n\t"
+      "smin %[result_11].4s, %[result_11].4s, v1.4s\n\t"
+      "smin %[result_00].4s, %[result_00].4s, v1.4s\n\t"
+      "smin %[result_01].4s, %[result_01].4s, v1.4s\n\t"
+
+      // Convert to float, multiply by the multiplier, and add the bias. Note
+      // that the float conversion instructions ("scvtf") are *very* slow and
+      // block the Neon pipeline. It is therefore important for optimal
+      // performance to interleave the float multiply and add instructions
+      // between the "scvtf" instructions.
```

Comment:
Great find! We might want to look into doing the same for the ruy bgemm kernels.